feat(curriculum): add email simulator workshop (#60935)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/workshop-email-simulator/6853f4ded9f84896ff182dd8.md

---
id: 6853f4ded9f84896ff182dd8
title: Step 19
challengeType: 20
dashedName: step-19
---

# --description--

> After the empty inbox check, add a print statement with the message `'\nYour Emails:'`. After that, iterate over all emails in the inbox using a `for` loop. Use enumeration to number the emails starting from `1`. Unpack the enumeration into `i` (the index) and `email` (the email object). Inside the loop, print each email using the `__str__` method you defined earlier, formatted with its number like this:

Reviewer: Directly mentioning `__str__` here can be misleading. People might think they should actually call `__str__`. I think so far we've never clearly stated how `__str__` is called under the hood. So this might need some explanations.

```suggestion
After the empty inbox check, print the message `'\nYour Emails:'`. After that, iterate over all emails in the inbox using a `for` loop. Use enumeration to number the emails starting from `1`. Use `i` (the index) and `email` (the email object) as the iteration variables.

Inside the loop, print an f-string made by the iteration index followed by a `.`, then a space and the string representation of the email object like this:
```

```md

> 1. [Formatted email summary from __str__ method]
> 2. [Formatted email summary from __str__ method]
> .
> .
> .
> n. [Formatted email summary from __str__ method]

Reviewer: I think a one-line example is fine here. Still I don't like to have all that text in the example.
```suggestion
[index]. [string representation of email object]
```

```

# --hints--

You should print the header `'\nYour Emails:'` outside the `if` statement.

```js
({ 
    test: () => 
    {
        assert(runPython(`_Node(_code).find_class("Inbox").find_function("list_emails").find_body()[1].is_equivalent("print('\\\\nYour Emails:')")`))
    }
})
```

> The enumeration should unpack into `i` and `email`.
> 
> ```js
> ({
>     test: () => 
>     {
>         assert(runPython(`_Node(_code).find_class("Inbox").find_function("list_emails").find_for_loops()[0].find_for_vars().is_equivalent("i,email")`))
>     }
> })
> ```
> 
> The enumeration should take `self.emails` and `start=1` as arguments.
> 
> ```js
> ({
>     test: () => 
>     {
>         assert(runPython(`_Node(_code).find_class("Inbox").find_function("list_emails").find_for_loops()[0].find_for_iter().is_equivalent("enumerate(self.emails, start=1)")`))
>     }
> })
> ```

Reviewer: It seems a bit convoluted

````suggestion
You should have a `for` loop that iterates over `enumerate(self.emails, 1)`.

```js
({
    test: () => 
    {
        runPython(`
_iter = _Node(_code).find_class("Inbox").find_function("list_emails").find_for_loops()[0].find_for_iter()
assert _iter.is_equivalent("enumerate(self.emails, start=1)") or _iter.is_equivalent("enumerate(self.emails, 1)")
`)
    }
})
```

Your `for` loop should use `i` and `email` to iterate over `enumerate(self.emails, 1)`.

```js
({
    test: () => 
    {
        assert(runPython(`_Node(_code).find_class("Inbox").find_function("list_emails").find_for_loops()[0].find_for_vars().is_equivalent("i, email")`))
    }
})
```
````

Inside the loop, you should print each email summary with a numbered prefix followed by the formatted string from the `__str__` method. Don't forget to add a `.` after the number.

```js
({
    test: () => 
    {

>         assert(runPython(`_Node(_code).find_class("Inbox").find_function("list_emails").find_for_loops()[0].find_body().is_equivalent("print(f'{i}. {str(email)}')")`))

Reviewer: Actually you don't need to use `str(email)` within the f-string. `print(f'{i}. {email}')` does the same. So we should remove it from the solution but allow both

Reviewer 2: @zairahira

Was this change applied?
